### PR TITLE
Add mimipenguin

### PR DESCRIPTION
**Pull request by bcoles**
Add `post/linux/gather/mimipenguin`

    This module searches process memory for system user credentials.

**work in progress / request for comments / request for testing**

This is a port of [mimipenguin](https://github.com/huntergregal/mimipenguin) to Metasploit.

### Requirements

* `root` privileges
* `strings` executable in `$PATH`
* `grep`executable (with support for `-E`)  in `$PATH`
* `gcore` or `dd` executable in `$PATH`

### Design

I've attempted to make the module efficient-ish by performing the majority of processing on the remote host. However, the following information is returned to Metasploit and stored in local memory for processing:

* all usernames and associated password hash
* all strings +/- 10 lines from each `needle` within each process of interest

No doubt I've made some unsafe assumptions; in particular:

* ~~The formatting and subsequent parsing of `crypt` hashes.~~
* If `grep` does not support the `-E` flag, things will likely go badly.
* If the partition for the dumpfile runs out of disk space, things will likely go badly.

### Dumping Memory

The module supports two techniques to dump process memory: `gcore` and `dd`.

* `gcore` is reliable but may not be present on all systems. `gcore` is the default technique used, when available, as it is faster than `dd`.

* `dd` is likely to be present on most systems. It also allows more granular control over which sections of memory are queried. This may be advantageous to improve efficiency in future implementations, but for now all process memory is dumped.

I had, but removed, a third option for `gdb`. Should anything go wrong while dumping memory with `gdb`, the process is likely to die a fiery death, which is more entertaining than it is desirable.

### Caveats

Most of the original mimipenguin functionality has been ported. The original also supported pulling HTTP Basic authentication credentials from Apache memory. I have not added this functionality. I do not intend to do so.

This module uses the same needles as the original `mimipenguin.py`. As such, it will fail to identify passwords in instances where mimipenguin also fails to identify passwords. This can easily be resolved by adding new needles as they're added to mimipenguin. Given that this is a shortcoming of the original project, of which this module is a port, I don't intend to improve this situation in this PR.

I've tested retrieving credentials from LightDM and gnome services successfully. Other services, such as vsftpd and ssh/sudo have not been tested. In theory, the same needles are used as in the original mimipenguin, so these should "just work".

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] Get a session
- [ ] `use post/linux/gather/mimipenguin`
- [ ] `set SESSION <id>`
- [ ] `run`
- [ ] You should receive credentials

## Documentation

I'll add documentation when the module is refined and tested.

### Scenarios

```
msf5 post(linux/gather/mimipenguin) > rexploit 
[*] Reloading module...

[*] Dumping credentials...
[+] [Gnome Keyring] Found credentials: user:password
[+] [LightDM] Found credentials: user:password
[+] Found 2 credentials

Credentials
===========

  Application    Username  Password
  -----------    --------  --------
  Gnome Keyring  user      password
  LightDM        user      password

[*] Credentials stored in /root/.msf4/loot/20180707025923_default_172.16.191.211_mimipenguin.csv_789726.txt
[*] Post module execution completed
msf5 post(linux/gather/mimipenguin) > cat /root/.msf4/loot/20180707025923_default_172.16.191.211_mimipenguin.csv_789726.txt
[*] exec: cat /root/.msf4/loot/20180707025923_default_172.16.191.211_mimipenguin.csv_789726.txt

Application,Username,Password
"Gnome Keyring","user","password"
"LightDM","user","password"
```

### CODE REVIEW

**Comment by huntergregal** [comment]
Hey this looks awesome bcoles! Really great work. I think this port is a great idea. I had some ideas a while back for potentially improving efficiency based on a more formal approach to locating the password in memory. I'll make a note to spend some time looking at the gnome-keyring structures in memory as a starting point and think about some other ways to improve the efficiency. 

If you have anything specific you'd like help on for this port don't hesitate to ask. I think MimiPenguin as a msf module would be it's best use case and i'm behind it 100%.

**Comment by huntergregal** [comment]
When targeting gnome-keyring-daemon: 3.18.3 installed in the default ubuntu, gdb attach the pid for gnome-keyring-daemon running as the target user and execute `x/s *(char**)0x6f7158+32 0x6f7158`. Following that pointer for me has resulted in the dump of the cleartext login password. 0x6f7158 appear to be a reference to the mmapped anonymous memory region for the egg-secure-memory. 

For efficiency in the module it may help to ptrace and follow these offsets and pointers to get the password instead of dumping all the memory. OR since we now know that the password is stored in an anonymously mapped memory region you can target only memory regions without a "path" as see in /proc/[pid]/maps

**Comment by bcoles** [comment]
@huntergregal 

If you can test and release a new stable a version of mimipenguin using the new technique, with new offsets, I'll happily look into porting it.

I'm regretting some of my design decisions - in particular the functionality to list processes - which I intend to fix. I also need to review the issue with `dd`. Once that's done, I can look at improving efficiency, or you're welcome to take over, and I can help out with porting as needed.

Regarding Ruby bindings for `ptrace`, while Metasploit is written in Ruby, Meterpreter is written in C, and the this module executes operating system commands on the target system. I've attempted to design the module in such a way that it will work using commands available on most systems. (caveat: `strings` and `grep` (with support for `-E`) executables are required).

Linux Meterpreter does not have methods to extract process memory. There's also no guarantee that `gdb` will be present on the target system. Additionally, executing within the Meterpreter process is somewhat different to executing a shell script. While it's possible to drop a shell script to disk and execute it, I would prefer to not do this, if possible.

I had support for `gdb` earlier (dumping all memory) but found that executing `gdb` is potentially problematic, as, should anything go wrong, the target process will likely die.

I also had issues with `dd`, which do not occur when executing mimipengiun from a shell, which I'm yet to investigate.

Given that mimipenguin uses the `dd` technique, and `dd` is capable of extracting memory segments via `/proc/<pid>/maps`, would it be possible to refine the needle search to a specific segment, for each process of interest, across multiple versions of the target software? if possible, I think this would be preferable to using `gdb`.

Ideally, in the long term, mimipenguin would be written in C as a Meterpreter extension. That will take a while to develop. In the mean time, it would be nice to get this functionality into Metasploit as a post module.

**Comment by bcoles** [comment]
I've cleaned up the process list parsing / retrieval in the previous commit.

As for the `dd` technique, Linux Meterpreter segfaults on some systems (CentOS and Fedora), however it works fine on a shell session. It also works on Linux Mint.

@huntergregal has released a new version of mimipenguin written in C. In the long term, this should be implemented as a Meterpreter extension.

Removing the delayed tag. Ready for testing.

**Comment by acammack-r7** [comment]
@bcoles Tried this from my Mac and since this relies on the native `crypt(3)` function on the msfconsole side I can't verify hashes from my Linux boxes (SHA512 in this case). Also, on my 14.04.4 test VM I needed to add another needle for `libdbus-1.so.3` to the GNOME keyring entry to successfully locate my password. I tried to find a pure Ruby version of `crypt(3)` to verify with, but [this is all I found](https://github.com/rubyworks/crypt3) and it does not work (looks like it's pulling the wrong digest algorithm).

**Comment by bcoles** [comment]
@acammack-r7 thanks for testing.

There's a few things going on here.

First, regarding the extra needle - does the original `mimipenguin.sh` or `mimipenguin.py` from the [beta-1.0 branch](https://github.com/huntergregal/mimipenguin/tree/beta-1.0) work on the target system? Mimipenguin should work on Ubuntu 14.04.4.

If not, I would prefer that new needles are pushed to the mimipenguin project, rather than straight to Metasploit. I'd rather keep this module as close to the original mimipenguin as possible, effectively outsourcing testing and maintenance upstream, so long as the mimipenguin project is alive.

Regarding `crypt`, that's problematic, as I don't own a Mac, so I might need some guidance or help with testing. Admittedly, my approach to `crypt` was naive.

In the interim, I offer the following workaround:

```ruby
unless RUBY_PLATFORM =~ /linux/
  fail_with Failure::BadConfig, 'Your operating system is bad and you should feel bad'
end
```

**Comment by acammack-r7** [comment]
Yep, neither of those extracts my password either, I'll file an issue over there.

Even that workaround isn't sure to work, since some Linux distros add a nonstandard crypt hash type `$2a$` for blowfish that's not in the upstream glibc. Maybe the `crypt3` gem can be rehabilitated to be less clever and work with modern Ruby?

**Comment by bcoles** [comment]
@acammack-r7 that might be my bad. Can you try mimipengiun beta-1.0 again before creating an issue? I recently changed, and subsequently reverted, an upstream change which turned out to suck.

**Comment by bcoles** [comment]
For those playing along at home, here's the relevant code:

```ruby
    @hashes.each do |h|
      hash = h[:hash]
      type = hash.split('$')[1]
      salt = hash.split('$')[2]
      if password.crypt("$#{type}$#{salt}").to_s.eql? hash
        users << h[:user]
      end
    end
```

And relevant `type` on Linux :

              ID  | Method
              ─────────────────────────────────────────────────────────
              1   | MD5
              2a  | Blowfish (not in mainline glibc; added in some
                  | Linux distributions)
              5   | SHA-256 (since glibc 2.7)
              6   | SHA-512 (since glibc 2.7)

**Comment by acammack-r7** [comment]
@bcoles Just tried again and still nothing (also the Python version crashes on `3.4.3`). I looked at the strings from the core of the keyring process, this GNOME keyring process for some reason just needs another needle.

**Comment by bcoles** [comment]
Processing will need to be performed on the target. Closing this in favor of implementing mimipenguin as a Meterpreter extension some time in the future.

**Comment by bcoles** [comment]
On second thought, even though it's horribly janky, and inefficient, and adds `python` as a dependency, it's probably better than nothing.

File: modules/post/linux/gather/mimipenguin.rb [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+  # Returns an array of users for which the specified password is valid
+  #
+  def check_valid_password(password)
+    users = []
+
+    @hashes.each do |h|
+      hash = h[:hash]
+      type = hash.split('$')[1]
+      salt = hash.split('$')[2]
+      python_code = "import crypt; import base64; print crypt.crypt(base64.b64decode('#{Rex::Text.encode_base64(password)}'), base64.b64decode('#{Rex::Text.encode_base64("$#{type}$#{salt}")}'))"
```
Comment (acammack-r7) [inline modules/post/linux/gather/mimipenguin.rb]: Python `crypt` also is not portable:
```
$ python3
>>> import crypt
>>> crypt.methods
[<crypt.METHOD_CRYPT>]
```

Comment (bcoles) [reply modules/post/linux/gather/mimipenguin.rb]: It's executed on the target.

Comment (acammack-r7) [reply modules/post/linux/gather/mimipenguin.rb]: Ah, so it is.

**Comment by bcoles** [comment]
> Processing will need to be performed on the target. Closing this in favor of implementing mimipenguin as a Meterpreter extension some time in the future.

New Mimipenguin implementation in #16688.

**Pull request closed without merging**